### PR TITLE
transmission: sent a warning and alias it to transmission_3

### PR DESCRIPTION
**Pull request by doronbehar**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
This looks like <code>transmission: rename **from** transmission_4</code>. And Transmission 4 contains at least one regression causing data loss, see the original PR: https://github.com/NixOS/nixpkgs/pull/215316#issuecomment-1485097961

**Comment by doronbehar** [comment]
> Transmission 4 contains at least one regression causing data loss

Why do you consider it as "data loss"? It's not files that were already downloaded are deleted from your storage, but rather it is the "torrent configuration" loss. I personally don't think it is such a bad bug to live with.

**Comment by jtojnar** [comment]
Yes, it loses torrent configuration in a way that downgrading does not restore. This is quite painful if you have large amount of affected torrents that you have configured. Thankfully, I had backups or I would be very annoyed.

**Comment by doronbehar** [comment]
Hmm I understand... Your usage of transmission is unique IMO. Perhaps we can add some release notes, with links to upstream bug reports? My personal opinion is that unless upstream released a completely broken package, a distribution shouldn't hold such significant updates from all users just because some users experience a bug that can barely be reproduced... I wasn't aware of this upstream release if I hadn't noticed it by chance by browsing Nixpkgs' pull requests.

BTW I don't understand what is the `ofborg-eval-darwin` error :/.

**Comment by doronbehar** [comment]
I'm attempting to fix the darwin build of transmission 4. I pushed several other improvements to both expressions, along with a release-notes line for the possible upstream bug experience.

**Comment by doronbehar** [comment]
@ofborg build transmission

**Comment by doronbehar** [comment]
The darwin build seems fixed.

**Comment by ThinkChaos** [comment]
I've been using 4.0.3 for a while, and lately ran into an issue where data was being downloaded, but not saved to disk. The torrents showed multiple GBs downloaded but still 0% progress. This was happening with all downloading torrents, even newly added ones, and was fixed by restarting Transmission.  
I haven't reported it upstream since 4.0.4 is out now and I'm not using it yet.

So even though I'm using v4 I'm against making it the default as in my experience it's not yet as stable as v3.

Your goal seems to be to raise awareness about v4. I can think of a couple ways to do that without risking people's data:
- add a release note mentioning v4 is packaged
- deprecate `transmission` and push users towards `transmission_3` or `transsmission_4`, similar to Python 2/3.

In both cases, we should explain why v4 isn't the default and link the PRs/upstream bug so users can make an informed decision.

**Comment by doronbehar** [comment]
Thanks @ThinkChaos for the ideas. I like them overall. As for:

> * deprecate `transmission` and push users towards `transmission_3` or `transsmission_4`, similar to Python 2/3.

Were you thinking of something like this in `aliases.nix`:

```nix
transmission = throw "Please instead of attribute `transmission`, use `transmission_3` or `transmission_4`";
```

vs:

```nix
transmission = lib.warn "The transmission attribute without version suffix is deprecated, please switch over to either transmission_4 or transmission_3. Note the release-notes mentioning upstream bugs." transmission_3;
```

Or simply (which would not make a difference for most users, unless aliases are disabled in their `nixpkgs.config`):

```nix
transmission = transmission_3;
```

I'd appreciate your vote on either of these options. Moreover, if you'd report that with 4.0.4, you don't experience the buggy behavior you observed with 4.0.3, that should affect this decision.

**Comment by kirillrdy** [comment]
> Thanks @ThinkChaos for the ideas. I like them overall. As for:
> 
> > * deprecate `transmission` and push users towards `transmission_3` or `transsmission_4`, similar to Python 2/3.
> 
> Were you thinking of something like this in `aliases.nix`:
> 
> ```nix
> transmission = throw "Please instead of attribute `transmission`, use `transmission_3` or `transmission_4`";
> ```
> 
> vs:
> 
> ```nix
> transmission = lib.warn "The transmission attribute without version suffix is deprecated, please switch over to either transmission_4 or transmission_3. Note the release-notes mentioning upstream bugs." transmission_3;
> ```
> 
> Or simply (which would not make a difference for most users, unless aliases are disabled in their `nixpkgs.config`):
> 
> ```nix
> transmission = transmission_3;
> ```
> 
> I'd appreciate your vote on either of these options. Moreover, if you'd report that with 4.0.4, you don't experience the buggy behavior you observed with 4.0.3, that should affect this decision.

sorry for noise
but what does 

`transmission = lib.warn` evaluate to ? null ? eg what is output when you try to build it ?

**Comment by doronbehar** [comment]
> `transmission = lib.warn` evaluate to ? null ? eg what is output when you try to build it ?

See: https://noogle.dev/?selected=%22lib.trivial.warn%22&term=%22warn%22

**Comment by kirillrdy** [comment]
> > `transmission = lib.warn` evaluate to ? null ? eg what is output when you try to build it ?
> 
> See: https://noogle.dev/?selected=%22lib.trivial.warn%22&term=%22warn%22

@doronbehar thank you for you response
@ThinkChaos I like your suggestion, I would be leading towards `throw` but that's just my opinion

**Comment by jtojnar** [comment]
> I wasn't aware of this upstream release if I hadn't noticed it by chance by browsing Nixpkgs' pull requests.

I would prefer `throw`ing over a silent migration to a broken version but I do not think there is much of a point. Nixpkgs is method for distributing software (which needs to be working to be useful) before being a mechanism for announcing latest versions of upstream projects. Also, eventually, we are going to remove `transmission_3` and then we would probably want to go back to using `transmission` attribute.

> Perhaps we can add some release notes, with links to upstream bug reports?

I do not think many people running unstable read release notes. Personally, I skim the commit log but it is really easy to miss something important there. As such, I would not consider a mention in release notes to be a sufficient precaution for bumping the `transmission` attribute to a version that eats data. 

There is a section for announcing new versions and Transmission 4 definitely deserves a mention (with a warning) but that just would not be enough for the bump IMO. 

> My personal opinion is that unless upstream released a completely broken package, a distribution shouldn't hold such significant updates from all users just because some users experience a bug that can barely be reproduced...

I would consider any data loss a serious breakage. I would not be against bumping the attribute if it was just the feature that was broken but it clears the torrent status so it is not possible to fix by downgrading once user ran 4.0. And the bug can be reproduced easily – just because it is a niche feature that does not affect many torrent files does not mean it does not seriously affect many peopleʼs workflows.

**Comment by ThinkChaos** [comment]
I would personally lean more towards `warn` for backward-compatibility reasons, and since it's similar to the established patern for module options (for instance `mkRenamedOptionModule`).  
That would also allow you to backport the change and reach people using stable.

> I do not think many people running unstable read release notes

I think you're right, personally I've started diffing the next release notes md file when updating my flake, works great :)

> I would not consider a mention in release notes to be a sufficient precaution for bumping the transmission attribute to a version that eats data.

I agree, I meant just as a way to say "`_4` exists, use it if you want".

**Comment by doronbehar** [comment]
I opened a PR that does everything done in this file besides the rename, so that the discussion here will stay focused: https://github.com/NixOS/nixpkgs/pull/259361 . Your review is welcome.

> There is a section for announcing new versions and Transmission 4 definitely deserves a mention (with a warning) but that just would not be enough for the bump IMO.

@jtojnar I feel we are in almost full agreement. Could you perhaps voice perhaps a more concrete opinion on the options suggested in [my previous comment above](https://github.com/NixOS/nixpkgs/pull/258058#issuecomment-1748390068)?

**Comment by doronbehar** [comment]
Haven't changed this PR fundamentally, still suggesting the same rename, just fixed the merge conflicts after #259361 .

**Comment by jtojnar** [comment]
Usually, we would first do alias without warning (evaluation checks did not like warnings during package set evaluation, not sure if we now run the checks with `allowAliases = false`), then switch to a throw a cycle later. The workflow is not ideal but :woman_shrugging: 

But I am questioning whether such action is desirable when eventually we are going to drop `transmission_3` eventually (hopefully sooner than later) and then we will want `transmission` to point to Transmission 4.

**Comment by doronbehar** [comment]
Just fixed evaluation from the rebase in the last force push. I think I'll come back to this later, and I'll give upstream a chance to respond to the issues presented to them, and suggest a solution afterwards.

**Comment by bb2020** [comment]
We also need these settings into [config file](https://github.com/NixOS/nixpkgs/blob/a31f1521f0b989e5f8eabaf4d237836d23a04706/nixos/modules/services/torrent/transmission.nix#L40).
`rename-partial-files = true;`
`lpd-enabled = false;`

LPD should be disabled by default as stated [here](https://trac.transmissionbt.com/ticket/3060#comment:39).
Strangely, partial files are not being renamed anymore with V4. Maybe this is related to https://github.com/NixOS/nixpkgs/pull/258058#issuecomment-1741416536?

**Comment by bb2020** [comment]
> @bb2020 are you trying to say that setting these may reduce the chance for the bugs reported in [#215316 (comment)](https://github.com/NixOS/nixpkgs/pull/215316#issuecomment-1485097961) to appear?

Normally incomplete files used to have `.part` extension. This new behavior leads to a massive confusion especially when `incomplete-dir-enabled` is disabled. Because of that we should override the default.

**Comment by jtojnar** [comment]
> Strangely, partial files are not being renamed anymore with V4. Maybe this is related to [#258058 (comment)](https://github.com/NixOS/nixpkgs/pull/258058#issuecomment-1741416536)?

I do not think that is related – that issue is with the torrent file itself, not the downloaded data.

**Comment by doronbehar** [comment]
Oh I didn't mean to close this, I just cleaned up my remote branches and deleted this old branch accidentally.

**Comment by bb2020** [comment]
Can you cherry pick [these changes](https://github.com/bb2020/nixpkgs/commit/b4ba876eaeae6cc5a0ef9bade1c43650a3d0baa6) into your PR? This should keep the older defaults.

**Comment by doronbehar** [comment]
> Can you cherry pick [these changes](https://github.com/bb2020/nixpkgs/commit/b4ba876eaeae6cc5a0ef9bade1c43650a3d0baa6) into your PR?

@bobby285271 I decided to change this PR a bit after the [discussion on discourse](https://discourse.nixos.org/t/how-to-introduce-a-breaking-version-bump/46746/1) and these changes maybe don't fit now. Perhaps you can submit a PR where the new settings you introduce change their default value based on the version of `cfg.package`?

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/torrent/transmission.nix
Comment (ambroisie) [inline nixos/modules/services/torrent/transmission.nix:179]: The documentation will make more sense with:

```suggestion
      package = mkPackageOption pkgs "transmission" {
        default = [ "transmission_3" ];
        example = "pkgs.transmission_4";
      };
```

Comment (doronbehar) [reply nixos/modules/services/torrent/transmission.nix]: > The documentation will make more sense with:

Thanks! Using a `default` seems weird to me, and I get the desired result with a single item instead:

image

**Review by ambroisie — APPROVED** [review]

**Review by kirillrdy — APPROVED** [review]

**Comment by bb2020** [comment]
> I decided to change this PR a bit after the [discussion on discourse](https://discourse.nixos.org/t/how-to-introduce-a-breaking-version-bump/46746/1) and these changes maybe don't fit now.

I see 4.00 still has some issues.

> Perhaps you can submit a PR where the new settings you introduce change their default value based on the version of `cfg.package`?

I think you should still cherry pick my changes.
They are the defaults for older version and would minimize confusion if someone switches between V3 and V4.

**Comment by jtojnar** [comment]
> I see 4.00 still has some issues.

Maybe we should create a thread in Discourse Announcements category like the following, and link it in the error message:

https://pad.lassul.us/lHdWr37LRxS-8gEHYWvfuA?both

(Feel free to edit)

> They are the defaults for older version and would minimize confusion if someone switches between V3 and V4.

I think this is not our responsibility. And doing so might instead cause confusion for people switching from other distribution to NixOS. At most, we can mention it in the announcement.

**Comment by AndersonTorres** [comment]
It looks like a nice use case for RFC 127 problem infrastructure!

**Review by jtojnar — DISMISSED** [review]
There is another reference to `pkgs.transmission` in `nixosTests.transmission`:

https://github.com/NixOS/nixpkgs/blob/c94ae9bb5dcb367937c99ceda87ff49e7b7b253e/nixos/tests/all-tests.nix#L980

**Comment by doronbehar** [comment]
Thanks for the ideas @jtojnar . I added both a notice to the upcoming release notes, based on [the text you wrote here](https://pad.lassul.us/lHdWr37LRxS-8gEHYWvfuA).

> > They are the defaults for older version and would minimize confusion if someone switches between V3 and V4.
> 
> I think this is not our responsibility. And doing so might instead cause confusion for people switching from other distribution to NixOS. At most, we can mention it in the announcement.

I agree. I'll write a comment at [the discourse thread](https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/) when this is merged.

File: pkgs/top-level/aliases.nix
Comment (jtojnar) [inline pkgs/top-level/aliases.nix:1299]: Maybe also mention the possible data loss and point to release notes?

Comment (doronbehar) [reply pkgs/top-level/aliases.nix:1299]: > Maybe also mention the possible data loss and point to release notes?

I was thinking about it but the warning became too long so I wasn't sure ... We don't care if it becomes too long? We do have the release notes after all...

Comment (jtojnar) [reply pkgs/top-level/aliases.nix:1299]: With the wording I am proposing above (edited), pointing to release notes makes it just 30% longer. Since we are ignoring the `nix-env` issue, I do not think we are that size constrained – people will just see the two lines of deprecation notice and then fix their config if they are annoyed.

Comment (doronbehar) [reply pkgs/top-level/aliases.nix]: Very well phrased. I used that.

**Comment by ThinkChaos** [comment]
I ran into another v4 bug that's technically data loss, sharing here for awareness: [Downloaded data not saved](https://github.com/transmission/transmission/issues/6796), might be nice to also add to the release notes.

**Comment by doronbehar** [comment]
> I ran into another v4 bug that's technically data loss, sharing here for awareness: [Downloaded data not saved](https://github.com/transmission/transmission/issues/6796), might be nice to also add to the release notes.

Linked to it in the release notes :+1:.

**Comment by doronbehar** [comment]
So.. @jtojnar do I get your final approval on this change?

**Comment by doronbehar** [comment]
Once again I fixed the merge conflicts. Since the last revision of this PR, I got only approvals and hadn't seen any objections to it. Hence I'll merge it myself in a day or two.

**Pull request merged**
